vsintegration/src/FSharp.Editor/CodeLens/CodeLens.fs

```diff
@@ -0,0 +1,799 @@
+                            let tags = line.GetAdornmentTags(self)
+                            match tags |> Seq.tryHead with
+                            | None -> ()
+                            | Some tag ->
+                                let uis = tag :?> StackPanel seq
+                                for ui in uis do
+                                    if ui |> isNull |> not then
+                                        ui.Visibility <- Visibility.Hidden
+                        with e -> logErrorf "Error in non visible lines iteration %A" e
+                for lineNumber in newVisibleLineNumbers do
+                    try
+                        let line = 
+                                let l = e.NewSnapshot.GetLineFromLineNumber(lineNumber)
+                                view.GetTextViewLineContainingBufferPosition(l.Start)
+                        let tags = line.GetAdornmentTags(self)
+                        match tags |> Seq.tryHead with
+                        | None -> ()
+                        | Some tag ->
+                            let uis = tag :?> StackPanel seq
+                            for ui in uis do
+                                if ui |> isNull |> not then
+                                    ui.Visibility <- Visibility.Visible
+                                    layoutUIElementOnLine view line.Extent ui
+                     with e -> logErrorf "Error in new visible lines iteration %A" e
+            // Save the new first and last visible lines for tracking
+            recentFirstVsblLineNmbr <- firstVisibleLineNumber
+            recentLastVsblLineNmbr <- lastVisibleLineNumber
+            // We can cancel existing stuff because the algorithm supports abortion without any data loss
+            layoutChangedCts.Cancel()
+            layoutChangedCts.Dispose()
+            layoutChangedCts <- new CancellationTokenSource()
+            asyncMaybe {
+                do! Async.SwitchToContext uiContext |> liftAsync
+                // Suspend 5 ms, insantly applying the layout to the adornment elements isn't needed 
+                // and would consume too much performance
+                do! Async.Sleep(5) |> liftAsync
+                let! view = view
+                let! layer = codeLensLayer
+                // Now relayout the existing adornments
+                if nonVisibleLineNumbers.Count > 0 || newVisibleLineNumbers.Count > 0 then
+                    let buffer = buffer.CurrentSnapshot
+                    for lineNumber in visibleLineNumbers do
+                        let line = 
+                            let l = buffer.GetLineFromLineNumber(lineNumber)
+                            view.GetTextViewLineContainingBufferPosition(l.Start)
+                        let tags = line.GetAdornmentTags(self)
+                        match tags |> Seq.tryHead with
+                        | None -> ()
+                        | Some tag ->
```

Comment:
`| Some (:? StackPanel seq as tag)`

Reply:
didn't know this one, awesome.

Reply:
But the `Stackpanel seq` must be wrapped with round brackets 😃

Reply:
But my suggestion won't throw on cast failures 😛